File: src/deadlock_build_sync/offline/beam_support.py
"""Count complete cores and assign new cores to frozen groups."""

from __future__ import annotations

from typing import TYPE_CHECKING

import numpy as np

from deadlock_build_sync.guide_generator import BEAM_SETTINGS
from deadlock_build_sync.purchase_windows import wilson_score_interval

if TYPE_CHECKING:
    from deadlock_build_sync.mechanics import ItemGraph

    from .discovery_data import HeroDiscoveryData


def wealth_state(value: float) -> int | None:
    if not np.isfinite(value) or value <= 0:
        return None
    return 0 if value < 0.9 else 2 if value > 1.1 else 1


def assign_core_group(
    core: frozenset[int], groups: dict[str, tuple[frozenset[int], ...]]
) -> str | None:
    existing = [group for group, members in groups.items() if core in members]
    if existing:
        return existing[0] if len(existing) == 1 else None
    eligible = [
        group
        for group, members in groups.items()
        if all(
            len(core & member) >= 2
            and len(core & member) / len(core | member)
            >= BEAM_SETTINGS["minimum_group_similarity"]
            for member in members
        )
    ]
    return eligible[0] if len(eligible) == 1 else None


class BeamOwnership:
    def __init__(self, values: HeroDiscoveryData, graph: ItemGraph, state: int) -> None:
        self.exact: dict[int, int] = {}
        self.expanded: dict[int, int] = {}
        self.eligible = 0
        for index, owned in enumerate(values.inventories):
            if (
                values.folds[index] != "discovery"
                or wealth_state(values.relative_wealth[index]) != state
            ):
                continue
            bit = 1 << index
            self.eligible |= bit
            for item in owned:
                self.exact[item] = self.exact.get(item, 0) | bit
                for component in (item, *graph.transitive_components(item)):
                    self.expanded[component] = self.expanded.get(component, 0) | bit
        self.cache: dict[tuple[tuple[int, ...], bool], int] = {}

    def count(self, core: tuple[int, ...], *, expanded: bool = False) -> int:
        key = tuple(sorted(core)), expanded
        if key not in self.cache:
            columns = self.expanded if expanded else self.exact
            owners = self.eligible
            for item in core:
                owners &= columns.get(item, 0)
            self.cache[key] = owners.bit_count()
        return self.cache[key]


def core_state_statistics(
    values: HeroDiscoveryData, core: tuple[int, ...], state: int
) -> dict[str, object]:
    result: dict[str, object] = {}
    required = set(core)
    ownership = np.asarray(
        [required.issubset(owned) for owned in values.inventories], dtype=bool
    )
    matched = np.asarray(
        [wealth_state(value) == state for value in values.relative_wealth], dtype=bool
    )
    for fold in ("discovery", "selection", "validation"):
        eligible = values.fold_mask(fold) & matched
        owners = eligible & ownership
        count, wins = int(owners.sum()), int(values.won[owners].sum())
        total = int(eligible.sum())
        lower, upper = wilson_score_interval(wins, count)
        result[fold] = {
            "owners": count,
            "wins": wins,
            "win_rate": wins / count if count else None,
            "lower_95": lower if count else None,
            "upper_95": upper if count else None,
            "hero_matches": total,
            "hero_win_rate": float(values.won[eligible].mean()) if total else None,
            "ownership_before_seconds": values.ownership_before_seconds,
        }
    return result

File: crates/deadlock-analysis/src/beam_support.rs
use std::collections::{BTreeMap, BTreeSet};

use deadlock_data::{Result, count_ratio};
use deadlock_guides::wilson_score_interval;
use deadlock_input::ItemGraph;
use serde_json::{Value, json};

use crate::discovery_data::DiscoveryData;

pub type CoreGroups = BTreeMap<String, Vec<BTreeSet<u64>>>;

#[derive(Debug)]
pub struct BeamOwnership {
    exact: BTreeMap<u64, Vec<u64>>,
    expanded: BTreeMap<u64, Vec<u64>>,
    eligible: Vec<u64>,
}

impl BeamOwnership {
    pub fn new(data: &DiscoveryData, graph: &ItemGraph, state: u8) -> Result<Self> {
        let words = data.rows.len().div_ceil(64);
        let mut result = Self {
            exact: BTreeMap::new(),
            expanded: BTreeMap::new(),
            eligible: vec![0; words],
        };
        for index in data
            .fold_rows("discovery")
            .filter(|index| wealth_state(data.rows[*index].relative_wealth) == Some(state))
        {
            let word = index / 64;
            let bit = 1 << (index % 64);
            result.eligible[word] |= bit;
            for item in &data.inventories[index] {
                result.exact.entry(*item).or_insert_with(|| vec![0; words])[word] |= bit;
                for component in std::iter::once(item).chain(graph.transitive_components(*item)?) {
                    result
                        .expanded
                        .entry(*component)
                        .or_insert_with(|| vec![0; words])[word] |= bit;
                }
            }
        }
        Ok(result)
    }

    pub fn count(&self, core: &[u64], expanded: bool) -> u64 {
        let columns = if expanded {
            &self.expanded
        } else {
            &self.exact
        };
        if core.iter().any(|item| !columns.contains_key(item)) {
            return 0;
        }
        self.eligible
            .iter()
            .enumerate()
            .map(|(index, eligible)| {
                u64::from(
                    core.iter()
                        .fold(*eligible, |bits, item| bits & columns[item][index])
                        .count_ones(),
                )
            })
            .sum()
    }
}

pub fn wealth_state(value: Option<f64>) -> Option<u8> {
    value
        .filter(|value| value.is_finite() && *value > 0.0)
        .map(|value| {
            if value < 0.9 {
                0
            } else if value > 1.1 {
                2
            } else {
                1
            }
        })
}

pub fn assign_core_group(core: &[u64], groups: &CoreGroups) -> Result<Option<String>> {
    let core = core.iter().copied().collect::<BTreeSet<_>>();
    let existing = groups
        .iter()
        .filter(|(_, members)| members.contains(&core))
        .map(|(group, _)| group)
        .collect::<Vec<_>>();
    if !existing.is_empty() {
        return Ok((existing.len() == 1).then(|| existing[0].clone()));
    }
    let mut eligible = Vec::new();
    for (group, members) in groups {
        let mut matched = true;
        for member in members {
            let shared = core.intersection(member).count();
            if shared < 2
                || count_ratio(
                    u64::try_from(shared)?,
                    u64::try_from(core.union(member).count())?,
                )? < 0.5
            {
                matched = false;
                break;
            }
        }
        if matched {
            eligible.push(group.clone());
        }
    }
    Ok((eligible.len() == 1).then(|| eligible[0].clone()))
}

pub fn state_statistics(data: &DiscoveryData, core: &[u64], state: u8) -> Result<Value> {
    let mut result = json!({});
    for fold in ["discovery", "selection", "validation"] {
        let eligible = data
            .fold_rows(fold)
            .filter(|index| wealth_state(data.rows[*index].relative_wealth) == Some(state))
            .collect::<Vec<_>>();
        let owners = eligible
            .iter()
            .filter(|index| {
                core.iter()
                    .all(|item| data.inventories[**index].contains(item))
            })
            .collect::<Vec<_>>();
        let count = u64::try_from(owners.len())?;
        let wins = u64::try_from(
            owners
                .iter()
                .filter(|index| data.rows[***index].won)
                .count(),
        )?;
        let total = u64::try_from(eligible.len())?;
        let hero_wins = u64::try_from(
            eligible
                .iter()
                .filter(|index| data.rows[**index].won)
                .count(),
        )?;
        let (lower, upper) = wilson_score_interval(wins, count, 1.96)?;
        result[fold] = json!({"owners":count,"wins":wins,"win_rate":if count>0{Some(count_ratio(wins,count)?)}else{None},
            "lower_95":if count>0{Some(lower)}else{None},"upper_95":if count>0{Some(upper)}else{None},
            "hero_matches":total,"hero_win_rate":if total>0{Some(count_ratio(hero_wins,total)?)}else{None},"ownership_before_seconds":1200});
    }
    Ok(result)
}
